design/metastore.md

```diff
@@ -0,0 +1,102 @@
+db/db1/table/tbl2 -> {tableName,dbName,owner,createTime,lastAccessTime,retention,sd,partitionKeys,...}
+db/db2/table/tbl1 -> {tableName,dbName,owner,createTime,lastAccessTime,retention,sd,partitionKeys,...}
+...
+```
+
+The data model that Hive Metastore provides enables more than the above. We will implement the missing entities as required by the data tools.
+
+
+### Thrift API
+
+Implementing the Hive Metastore 2.3.x will be the starting point of lakeFS's Metastore.
+Spark works with Hive Metastore version 2.3.9 (https://spark.apache.org/docs/latest/sql-data-sources-hive-tables.html), but can also work with version 3.x.
+Trino can work with Hive 2 and 3, while having specific capabilities enabled for version 3.
+Must of the Hive Metastore Thrift interface kept backward compatibility v2.3.x and v3.x interfaces are almost identical, where version 3 adds Catalog and Workload management metadata.
+
+Code generation should be considered. Having generated bride that will transfer each Thrift API call to our OpenAPI. And/or generate code for each entity described in the Thrift interface can reduce the code in order serialize, map and filter the data model.
+
+Thrift API communicate over a socket, the first step will enable non-secure communication over an additional port. A secure communication can be handled by a external TCP load-balancer first, later the proxy can supply this capability out of the box.
+
+
+Hooks - The Thrift interface include a set of API ends with '_environment_context', which enables the caller passing additional information to the hooks. The hooks mechanism in Hive Metastore is a Java interface implementation that the metastore load and calls on specific operations. We will not support this mechanism at first, but we can have a reference implementation of what to expect in case of adding it to our hooks mechanism.
+
+
+### Discovery
+
+Metastore implementation provides a reference implementation we can test and explore. It still needs to be discovered how to extend functionality to enable versioning.
+
+Finding out more about the following items will help us evaluate our design options:
+
+1. Passing lakeFS's reference/branch information from the metastore client. Explore ways to pass repository/branch/ref information from the client to our metastore.
+1. Side by side - very similar to the current work on side-by-side work with lakeFS and the current object-storage, the user like to run the same application with minimum changes, while enabling some of the data to live inside lakeFS. What needs to change or requirements in order to have the same in the metadata level? Do we need to sync information, manage part of the data and enable mapping on the client side? Can we replace the client side implementation? do we need to spin different endpoints to serve different repositories, branches in lakeFS for parts of the metadata? fallback to existing metastore?
```

Comment:
I think an example is in order here, not sure what it means to have a side-by-side here.

> ... enable mapping on the client side? Can we replace the client side implementation?

Client side of what?